src/synchronization_c.c4

```diff
@@ -164,3 +304,100 @@ SHMEM_BIND_C_SYNC(`SHMEM_DEF_WAIT_UNTIL')
+    {                                                                                          \
+        int cmpret, ncompleted = 0;                                                            \
+        SHMEM_ERR_CHECK_INITIALIZED();                                                         \
+        SHMEM_ERR_CHECK_SYMMETRIC(vars, sizeof(TYPE));                                         \
+        SHMEM_ERR_CHECK_CMP_OP(cond);                                                          \
+                                                                                               \
+        if (nelems == 0) return 1;                                                             \
+                                                                                               \
+        for (int i = 0; i < nelems; i++) {                                                     \
+            COMP(cond, vars[i], value, cmpret);                                                \
+            if (cmpret) ncompleted++;                                                          \
+        }                                                                                      \
+        if (ncompleted == nelems) {                                                            \
+            shmem_internal_membar_load();                                                      \
+            return 1;                                                                          \
+        } else {                                                                               \
+            return 0;                                                                          \
+        }                                                                                      \
+    }
+
+SHMEM_BIND_C_SYNC(`SHMEM_DEF_TEST_ALL')
+
+
+#define SHMEM_DEF_TEST_ANY(STYPE,TYPE)                                                         \
+    size_t SHMEM_FUNCTION_ATTRIBUTES                                                           \
+    shmemx_##STYPE##_test_any(TYPE *vars, size_t nelems, int * restrict status,                \
+                              int cond, TYPE value)                                            \
+    {                                                                                          \
+        SHMEM_ERR_CHECK_INITIALIZED();                                                         \
+        SHMEM_ERR_CHECK_SYMMETRIC(vars, sizeof(TYPE));                                         \
+        SHMEM_ERR_CHECK_CMP_OP(cond);                                                          \
+                                                                                               \
+        size_t found_idx = SIZE_MAX, i = 0;                                                    \
+        unsigned int rand_seed = shmem_internal_my_pe;                                         \
+        size_t start_idx = rand_r(&rand_seed) / (RAND_MAX + 1.0) * nelems;                     \
```

Comment:
It would also be good to have a unit test that verifies repeated calls to test-any (and also test-some) eventually flag every ivar.  You don't need shmem communication for this, just initialize an ivars array to all `1` and then:
```
while (nfound < nelems) {
  int idx = shmemx_test_any(ivars, nelems, NULL, SHMEM_CMP_EQ, 1);
  if (found[idx] == 0) {
    found[idx] = 1;
    ++nfound;
  }
}
```

Reply:
Whoops!  😮 
Yup, I'll fix that.  And test it this time...